File: implementations/java/ff3-cli/src/main/java/dev/horizondigital/ff3/cli/FF3CLI.java
package dev.horizondigital.ff3.cli;

import dev.horizondigital.ff3.FF3API;
import dev.horizondigital.ff3.FF3Alphabets;

@SuppressWarnings("java:S106") // System.out is appropriate for CLI application output
public class FF3CLI {

    public static void main(String[] args) {
        try {
            Options opts = parseArgs(args);

            if (opts.help) {
                showUsage();
                return;
            }

            // Validate required arguments
            if (opts.encryptText == null && opts.decryptText == null) {
                System.err.println("Error: Either --encrypt or --decrypt must be specified");
                showUsage();
                System.exit(1);
            }

            if (opts.keyHex == null || opts.tweakHex == null) {
                System.err.println("Error: Key (-k) and tweak (-t) are required");
                showUsage();
                System.exit(1);
            }

            // Parse key and tweak
            byte[] key = FF3API.hexToBytes(opts.keyHex);
            byte[] tweak = FF3API.hexToBytes(opts.tweakHex);

            // Validate key length
            if (key.length != 16 && key.length != 24 && key.length != 32) {
                System.err.printf("Error: Key must be 16, 24, or 32 bytes (32, 48, or 64 hex chars), got %d bytes%n", key.length);
                System.exit(1);
            }

            // Validate tweak length
            if (tweak.length != 8) {
                System.err.printf("Error: Tweak must be 8 bytes (16 hex chars), got %d bytes%n", tweak.length);
                System.exit(1);
            }

            // Create alphabet
            String alphabet = createAlphabet(opts.alphabetType, opts.customCharset);

            // Create cipher
            FF3API ff3 = new FF3API(key, tweak, alphabet);

            // Encrypt or decrypt
            if (opts.encryptText != null) {
                String result = ff3.encrypt(opts.encryptText);
                System.out.println(result);
            } else if (opts.decryptText != null) {
                String result = ff3.decrypt(opts.decryptText);
                System.out.println(result);
            }

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }

    private static Options parseArgs(String[] args) {
        Options opts = new Options();

        for (int i = 0; i < args.length; i++) {
            String arg = args[i];
            switch (arg) {
                case "-h":
                case "--help":
                    opts.help = true;
                    return opts;
                case "-e":
                case "--encrypt":
                    if (i + 1 >= args.length) {
                        throw new IllegalArgumentException("Missing value for " + arg);
                    }
                    opts.encryptText = args[++i];
                    break;
                case "-d":
                case "--decrypt":
                    if (i + 1 >= args.length) {
                        throw new IllegalArgumentException("Missing value for " + arg);
                    }
                    opts.decryptText = args[++i];
                    break;
                case "-k":
                case "--key":
                    if (i + 1 >= args.length) {
                        throw new IllegalArgumentException("Missing value for " + arg);
                    }
                    opts.keyHex = args[++i];
                    break;
                case "-t":
                case "--tweak":
                    if (i + 1 >= args.length) {
                        throw new IllegalArgumentException("Missing value for " + arg);
                    }
                    opts.tweakHex = args[++i];
                    break;
                case "-a":
                case "--alphabet":
                    if (i + 1 >= args.length) {
                        throw new IllegalArgumentException("Missing value for " + arg);
                    }
                    opts.alphabetType = args[++i];
                    break;
                case "-c":
                case "--custom":
                    if (i + 1 >= args.length) {
                        throw new IllegalArgumentException("Missing value for " + arg);
                    }
                    opts.customCharset = args[++i];
                    break;
                default:
                    throw new IllegalArgumentException("Unknown option: " + arg);
            }
        }

        return opts;
    }

    private static String createAlphabet(String alphabetType, String customCharset) {
        if (customCharset != null && !customCharset.isEmpty()) {
            return customCharset;
        }

        if (alphabetType == null) {
            alphabetType = "digits";
        }

        switch (alphabetType.toLowerCase()) {
            case "digits":
                return FF3Alphabets.DIGITS;
            case "hex":
            case "hex-lower":
                return FF3Alphabets.HEX_LOWER;
            case "hex-upper":
                return FF3Alphabets.HEX_UPPER;
            case "base36":
            case "base36-lower":
                return FF3Alphabets.BASE36_LOWER;
            case "base36-upper":
                return FF3Alphabets.BASE36_UPPER;
            case "base62":
                return FF3Alphabets.BASE62;
            default:
                throw new IllegalArgumentException("Unknown alphabet type: " + alphabetType);
        }
    }

    private static void showUsage() {
        System.out.println("FF3 CLI - Format Preserving Encryption");
        System.out.println();
        System.out.println("Usage: ff3-cli [OPTIONS]");
        System.out.println();
        System.out.println("Options:");
        System.out.println("  -e, --encrypt TEXT      Encrypt the given text");
        System.out.println("  -d, --decrypt TEXT      Decrypt the given text");
        System.out.println("  -k, --key HEX           AES key in hex format (32/48/64 hex chars)");
        System.out.println("  -t, --tweak HEX         Tweak in hex format (16 hex chars)");
        System.out.println("  -a, --alphabet TYPE     Alphabet type:");
        System.out.println("                            digits (default)");
        System.out.println("                            hex-lower");
        System.out.println("                            hex-upper");
        System.out.println("                            base36-lower");
        System.out.println("                            base36-upper");
        System.out.println("                            base62");
        System.out.println("  -c, --custom CHARSET    Custom alphabet charset");
        System.out.println("  -h, --help              Show this help message");
        System.out.println();
        System.out.println("Examples:");
        System.out.println("  ff3-cli -e \"1234567890\" -k EF4359D8D580AA4F7F036D6F04FC6A94 -t D8E7920AFA330A73");
        System.out.println("  ff3-cli -d \"7501889140\" -k EF4359D8D580AA4F7F036D6F04FC6A94 -t D8E7920AFA330A73");
        System.out.println();
    }

    private static class Options {
        boolean help = false;
        String encryptText = null;
        String decryptText = null;
        String keyHex = null;
        String tweakHex = null;
        String alphabetType = "digits";
        String customCharset = null;
    }
}

File: implementations/erlang/apps/ff3_cli/src/ff3_cli.erl
%% FF3 CLI Main Module

-module(ff3_cli).
-export([main/1]).

main(Args) ->
    case parse_args(Args, #{}) of
        {ok, Opts} ->
            run(Opts);
        {error, help} ->
            show_usage(),
            erlang:halt(0);
        {error, Reason} ->
            io:format("Error: ~s~n", [Reason]),
            erlang:halt(1)
    end.

show_usage() ->
    io:format("FF3 CLI - Format Preserving Encryption~n~n"),
    io:format("Usage: ff3-cli [OPTIONS]~n~n"),
    io:format("Options:~n"),
    io:format("  -e, --encrypt TEXT      Encrypt the given text~n"),
    io:format("  -d, --decrypt TEXT      Decrypt the given text~n"),
    io:format("  -k, --key HEX           AES key in hex format (32/48/64 hex chars)~n"),
    io:format("  -t, --tweak HEX         Tweak in hex format (16 hex chars)~n"),
    io:format("  -a, --alphabet TYPE     Alphabet type:~n"),
    io:format("                            digits (default)~n"),
    io:format("                            hex-lower~n"),
    io:format("                            hex-upper~n"),
    io:format("                            base36-lower~n"),
    io:format("                            base36-upper~n"),
    io:format("                            base62~n"),
    io:format("  -c, --custom CHARSET    Custom alphabet charset~n"),
    io:format("  -h, --help              Show this help message~n~n"),
    io:format("Examples:~n"),
    io:format("  ff3-cli -e \"1234567890\" -k EF4359D8D580AA4F7F036D6F04FC6A94 -t D8E7920AFA330A73~n"),
    io:format("  ff3-cli -d \"7501889140\" -k EF4359D8D580AA4F7F036D6F04FC6A94 -t D8E7920AFA330A73~n~n").

parse_args([], Opts) ->
    validate_opts(Opts);
parse_args(["-h"|_], _Opts) ->
    {error, help};
parse_args(["--help"|_], _Opts) ->
    {error, help};
parse_args(["-e", Text|Rest], Opts) ->
    parse_args(Rest, Opts#{encrypt => Text});
parse_args(["--encrypt", Text|Rest], Opts) ->
    parse_args(Rest, Opts#{encrypt => Text});
parse_args(["-d", Text|Rest], Opts) ->
    parse_args(Rest, Opts#{decrypt => Text});
parse_args(["--decrypt", Text|Rest], Opts) ->
    parse_args(Rest, Opts#{decrypt => Text});
parse_args(["-k", KeyHex|Rest], Opts) ->
    parse_args(Rest, Opts#{key => KeyHex});
parse_args(["--key", KeyHex|Rest], Opts) ->
    parse_args(Rest, Opts#{key => KeyHex});
parse_args(["-t", TweakHex|Rest], Opts) ->
    parse_args(Rest, Opts#{tweak => TweakHex});
parse_args(["--tweak", TweakHex|Rest], Opts) ->
    parse_args(Rest, Opts#{tweak => TweakHex});
parse_args(["-a", Type|Rest], Opts) ->
    parse_args(Rest, Opts#{alphabet => Type});
parse_args(["--alphabet", Type|Rest], Opts) ->
    parse_args(Rest, Opts#{alphabet => Type});
parse_args(["-c", Charset|Rest], Opts) ->
    parse_args(Rest, Opts#{custom => Charset});
parse_args(["--custom", Charset|Rest], Opts) ->
    parse_args(Rest, Opts#{custom => Charset});
parse_args([Unknown|_], _Opts) ->
    {error, io_lib:format("Unknown option: ~s", [Unknown])}.

validate_opts(Opts) ->
    case {maps:is_key(encrypt, Opts), maps:is_key(decrypt, Opts)} of
        {false, false} ->
            {error, "Either --encrypt or --decrypt must be specified"};
        _ ->
            case {maps:is_key(key, Opts), maps:is_key(tweak, Opts)} of
                {false, _} ->
                    {error, "Key (-k) is required"};
                {_, false} ->
                    {error, "Tweak (-t) is required"};
                _ ->
                    {ok, Opts}
            end
    end.

run(Opts) ->
    try
        KeyHex = maps:get(key, Opts),
        TweakHex = maps:get(tweak, Opts),
        AlphabetType = maps:get(alphabet, Opts, "digits"),
        CustomCharset = maps:get(custom, Opts, undefined),

        Key = ff3_api:hex_to_bytes(KeyHex),
        Tweak = ff3_api:hex_to_bytes(TweakHex),

        %% Validate key length
        KeyLen = byte_size(Key),
        case KeyLen of
            16 -> ok;
            24 -> ok;
            32 -> ok;
            _ -> throw({invalid_key_length, KeyLen})
        end,

        %% Validate tweak length
        case byte_size(Tweak) of
            8 -> ok;
            _ -> throw({invalid_tweak_length, byte_size(Tweak)})
        end,

        Alphabet = create_alphabet(AlphabetType, CustomCharset),

        case {maps:get(encrypt, Opts, undefined), maps:get(decrypt, Opts, undefined)} of
            {undefined, Ciphertext} ->
                Plaintext = ff3_api:decrypt(Ciphertext, Key, Tweak, Alphabet),
                io:format("~s~n", [Plaintext]);
            {Plaintext, _} ->
                Ciphertext = ff3_api:encrypt(Plaintext, Key, Tweak, Alphabet),
                io:format("~s~n", [Ciphertext])
        end
    catch
        _:Error ->
            io:format("Error: ~p~n", [Error]),
            erlang:halt(1)
    end.

create_alphabet(_Type, CustomCharset) when CustomCharset =/= undefined ->
    CustomCharset;
create_alphabet("digits", _) ->
    ff3_alphabets:digits_alphabet();
create_alphabet("hex-lower", _) ->
    ff3_alphabets:hex_lower_alphabet();
create_alphabet("hex-upper", _) ->
    ff3_alphabets:hex_upper_alphabet();
create_alphabet("base36-lower", _) ->
    ff3_alphabets:base36_lower_alphabet();
create_alphabet("base36-upper", _) ->
    ff3_alphabets:base36_upper_alphabet();
create_alphabet("base62", _) ->
    ff3_alphabets:base62_alphabet();
create_alphabet(Other, _) ->
    throw({unknown_alphabet, Other}).
